File: solved/search-a-2d-matrix.ts
/**
 * https://leetcode.com/problems/search-a-2d-matrix/description/
 */
export { searchMatrix, binarySearchRow, binarySearchCol };

function searchMatrix(matrix: number[][], target: number): boolean {
  const rowIndex = binarySearchCol(matrix, target);
  if (rowIndex === -1) return false;
  return binarySearchRow(matrix[rowIndex], target);
}

function binarySearchCol(matrix: number[][], target: number): number {
  const rowLastIndex = matrix.length - 1;

  for (let row = 0; row <= rowLastIndex - 1; row++) {
    if (matrix[row][0] <= target && target < matrix[row + 1][0]) return row;
  }

  return rowLastIndex;
}

function binarySearchRow(row: number[], target: number): boolean {
  function bs(start: number, end: number): boolean {
    if (row[start] === target) return true;
    if (row[end] === target) return true;
    if (start >= end) return false;

    const mid = Math.floor((start + end) / 2);

    if (target <= row[mid]) return bs(start, mid);
    else return bs(mid + 1, end);
  }

  return bs(0, row.length);
}

File: returned/python/today5/search_a_2_d_matrix.py
# https://leetcode.com/problems/search-a-2d-matrix/description/
from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row_count, col_count = len(matrix), len(matrix[0])
        top, bottom = 0, row_count-1
        row_index = -1
        while top <= bottom:
            mid = (top + bottom) // 2
            if matrix[mid][0] <= target <= matrix[mid][col_count-1]:
                row_index = mid
                break
            elif target < matrix[mid][0]:
                bottom = mid - 1
            else:
                top = mid + 1
        
        if row_index == -1:
            return False
        
        left, right = 0, col_count-1
        while left <= right:
            mid = (left + right) // 2
            if target == matrix[row_index][mid]:
                return True
            elif target < matrix[row_index][mid]:
                right = mid - 1
            else:
                left = mid + 1
                
        return False
